src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/AttendanceService.java

```diff
@@ -276,32 +276,69 @@ private Boolean isAttendanceLocked(final Attendance attendance) {
                 .findByPrisonIdAndEventDateBetweenAndPeriodInAndAbsentReason(prisonId, fromDate, endDate, periods, absentReason);
 
         return attendances.stream()
-                .map(this::toAttendanceDto)
-                .filter(attendance -> offenderDetails.stream().anyMatch(findAttendance(attendance)))
-                .map(attendanceDto -> attendanceDto
-                        .toBuilder()
-                        .cellLocation(offenderDetails.stream().filter(findAttendance(attendanceDto)).findFirst().map(OffenderDetails::getCellLocation).orElse(""))
-                        .build())
+                .filter(attendance -> offenderDetails.stream().anyMatch(findAttendance(attendance.getBookingId(), attendance.getEventId(), attendance.getPeriod())))
+                .map(attendance -> {
+                    final var details = offenderDetails.stream()
+                            .filter(findAttendance(attendance.getBookingId(), attendance.getEventId(), attendance.getPeriod())).findFirst()
+                            .orElseThrow();
+
+                    return toAbsenceDto(details, attendance);
+                })
                 .collect(Collectors.toSet());
     }
 
-    private Predicate<? super OffenderDetails> findAttendance(final AttendanceDto attendance) {
-        return offender -> offender.getBookingId() == attendance.getBookingId() && offender.getEventId() == attendance.getEventId() && TimePeriod.valueOf(offender.getTimeSlot()) == attendance.getPeriod();
+    private OffenderDetails offenderDetailsWithPeriod(final OffenderDetails details, final TimePeriod period) {
+        return details.copy(
+                details.getBookingId(),
+                details.getOffenderNo(),
+                details.getEventId(),
+                details.getCellLocation(),
+                details.getEventDate(),
+                period.toString(),
+                details.getFirstName(),
+                details.getLastName(),
+                details.getComment()
+        );
+    }
+
+    private AbsenceDto toAbsenceDto(final OffenderDetails details, final Attendance attendance) {
+        return new AbsenceDto(
+                attendance.getId(),
+                attendance.getBookingId(),
+                details.getOffenderNo(),
+                attendance.getEventId(),
+                attendance.getEventLocationId(),
+                attendance.getEventDate(),
+                TimePeriod.valueOf(details.getTimeSlot()),
+                attendance.getAbsentReason(),
+                details.getComment(),
+                attendance.getComments(),
+                details.getCellLocation(),
+                details.getFirstName(),
+                details.getLastName()
+        );
+    }
+
+    private Predicate<? super OffenderDetails> findAttendance(final Long bookingId, final Long eventId, final TimePeriod period) {
+        return offender -> offender.getBookingId().equals(bookingId) &&
```

Comment:
Probably minor, but it looks like OffenderDetails has a nullable bookingId? Could this raise a NullPointerException if it's not there? Same with the eventId below.

Reply:
Good spot, bookingId is fine, however eventId, at least in dev is sometimes null due to crappy data. I'll fix it anyways just to be safe.